SPM support

DevAndArtist opened the pull request:
This PR completes #309.

I had to dance a little with some project files as the framework contains mixed languages (Swift + Objective-C). Therefore the package contains a standalone lib for the hex2bin converter which the main lib `DFULibrary` depends on. 

Cocoapods and Carthage both do not need that lib, that's why we use `#if canImport(...)` which is only required for SP. 

The header file is also moved into its own `include` directory or otherwise SPM would complain about the missing directory.

Also note that I didn't called the library `iOSDFULibrary` because we do support macOS and it does not make sense to call it that way.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/NordicSemiconductor/IOS-Pods-DFU-Library?pullRequest=310) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/NordicSemiconductor/IOS-Pods-DFU-Library?pullRequest=310) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/NordicSemiconductor/IOS-Pods-DFU-Library?pullRequest=310) it.</sub>

DevAndArtist commented:
cc @philips77

DevAndArtist commented:
To try out building the project with SPM you need Xcode 11. It's enough to double click on `Package.swift`. Xcode will open not just the file but an entire project structure like that. You can select the scheme you want and just try building.

![Screenshot 2019-06-12 at 11 50 43](https://user-images.githubusercontent.com/8276230/59341779-beb07180-8d08-11e9-9aa9-73ef4a416eee.png)

DevAndArtist commented:
@philips77 an optimisation would be if you could re-write the Objective-C code in to Swift, then we could remove the `Hex2BinConverter` and keep it as an implementation detail as intended. However right now it leaks to the consumer.

philips77 commented:
Many, many thanks! We will review it after drinking the champagne!
Rewriting the Hex2BinConverter is a good idea. Let me do this first before merging this PR so we could have clean SPM setup.

philips77 commented:
Regarding the iOS in the name, this is mainly to distinguish from the Android lib, which has Android in the name. macOS support was added... accidentally, as it just works. But I support the idea for publishing as DFULibrary.

DevAndArtist commented:
@philips77 by the way the package also compiles just fine for tvOS and watchOS.

DevAndArtist commented:
@philips77 now that we have the chance, we could also use a better name then something with a `Library` suffix. Maybe `NordicDFU`?

dinesharjani commented on iOSDFULibrary/Classes/Utilities/ZipArchive.swift:
```diff
@@ -21,6 +21,7 @@
 */
 
 import ZIPFoundation
+import Foundation
```
@DevAndArtist this entire diff looks amazing! Thank so much!
Could you explain why you added the `Foundation` import here? It doesn't seem like we need it if there are no other changes to this file.

dinesharjani replied:
This is the only "blocking" comment.

DevAndArtist replied:
In an Xcode framework the import is implicitly, but in SPM you need to be a little bit more explicit what you want to use and where the compiler has to lookup the code. SPM complained that this particular file used unknown files such as `Data`. Therefore I added an explicit import of `Foundation`.

DevAndArtist replied:
@dinesharjani it wasn't much. Again as I mentioned above I would appreciate if we could call the library more like other Swift libraries. `NordicDFU` seems like a good candidate. Furthermore we could consider to split the package to provide two different products.

- `NordicDFU` without a dependency to `ZIPFoundation` (for example in our use of `iOSDFULibrary` we don't need the archiving part at all and bundling an extra library with our product seems unnecessary)
- `NordicDFUExtension` a convenient extension of the library that adds support for zip archives.

DevAndArtist replied:
But we can do this inclemently if you want. ;)

dinesharjani commented on iOSDFULibrary/Classes/Utilities/Streams/DFUStreamHex.swift:
```diff
@@ -20,6 +20,9 @@
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
 import Foundation
+#if canImport(Hex2BinConverter)
```
Thank you for explaining why this change is needed.

dinesharjani commented on README.md:
```diff
@@ -45,6 +45,24 @@
 you may now copy all those files to your project and use the library, additionally, carthade also builds **\*.dsym** files 
 if you need to resymbolicate crash logs. you may want to keep those files bundled with your builds for future use.
 
+**For Swift Package Manager:**
```
Again, thank you for making such a complete diff

philips77 commented:
Hi, have a look at https://github.com/NordicSemiconductor/IOS-Pods-DFU-Library/blob/feature/swift-hex-converter/iOSDFULibrary/Classes/Utilities/HexToBinConverter/IntelHex2BinConverter.swift
(branch **feature/swift-hex-converter**, commits:
https://github.com/NordicSemiconductor/IOS-Pods-DFU-Library/commits/feature/swift-hex-converter)

DevAndArtist commented:
@philips77 I only had a very quick glance at the code. I don't really have time atm. to evaluate the correctness of the code according to its specifications, but for that you build some tests which is usually a good thing. :) 

If you're confident with the changes feel free to merge it into my branch (you have the rights to do so as a maintainer of the original project).

After that is done, I'll clean up `Package.swift` file.

---

Should I call the package and lib `NordicDFU` then?

philips77 commented:
I'll merge tomorrow. Yes, let's call them `NordicDFU`.

philips77 commented:
Ok, I managed to merge latest commits into your branch, but you need to review it. I discarded all changes in `Pods.xcodeproj/project.pbxproj`, as they were related to `IntelHex2BinConverter.h`, but I might have missed something. I had to rebase the branch.
Could you check if it still works and fix if needed? Also, as discussed, let's set the name to `NordicDFU`.

DevAndArtist commented:
@philips77 I updated the project. I also added the tests target to the package and fixed some compile-time errors when targeting `watchOS` (except for one package issue related to `watchOS` and tests - see comment in `Package.swift`). 

The tests are also re-routed to use the URL of the test file instead grabbing the resources from the bundle. This change enables us to run them from both SPM and from `.xcworkspace`.

philips77 commented:
Amazing! Thanks again for all the work.
Is there anything else blocking from releasing? Is there a need for a tag after releasing?

DevAndArtist commented:
> Is there anything else blocking from releasing?

I don't think so. In the future one could clean up the project structure and `CocoaPod/Carthage` module names, but that's not relevant now.

> Is there a need for a tag after releasing?

No sure, but I would guess that it's needed.
